### DISCUSSION TITLE
Troubleshooting "no space left on device"

### DISCUSSION THREAD

**Discussion by vito** [concourse/concourse#5463] [Help & Support]
This discussion is centered on a common error: `no space left on device`. This is an error that Concourse can try to prevent, but never truly "fix" - if you try to fetch a 100GB resource on a worker with 10GB of space, there's nothing that can be done to prevent the error.

Nevertheless, users - especially new ones - will likely run into this error and Google it, so this Discussion is here to provide troubleshooting tips and extensive information about how Concourse uses disk space so that we can all more easily understand how to size worker disks appropriately, have a better idea of what volume driver to use, and finally to know how to troubleshoot an error to see if there's really a bug involved or if you just need more disk space.

Before replying to this discussion, reporting a bug, or opening a new discussion, please skim this document and read through the troubleshooting tips so that we can get a head start with as much information as possible. Troubleshooting this issue is very difficult from afar.

Thank you! :slightly_smiling_face:

## How volumes are used by Concourse

All data on disk is stored in "volumes" (directories) in the worker's `--work-dir`. These volumes are mounted to containers.

Each container has an *image* - a volume containing the root filesystem of the container. For a task, this volume may be specified as a resource to fetch using [`image_resource:`](https://concourse-ci.org/tasks.html#schema.task.image_resource), or it may be specified on the `task` step using [`image:`](https://concourse-ci.org/jobs.html#schema.step.task-step.image). For a resource, this volume is either configured as the type's image under [`resource_types:`](https://concourse-ci.org/resource-types.html#schema.resource_type), or is a volume that is preconfigured on the worker as a "base" resource type.

A `task` step container has a volume for each input and a volume for each output, which the task writes to. A resource `get` step container has no input volumes and fetches the resource version's bits to an output volume. A resource `put` step container has a volume for each input, and has no output volumes. A resource `check` container has no input or output volumes.

Concourse guarantees that each container runs in a clean and isolated environment so that it's impossible to pollute data in a concurrent container and/or subsequent containers in the current build or later builds. Many "old-school" CI systems don't have the same guarantee, instead running builds in a shared workspace, typically without containers at all. This is obviously faster in the sense that it's doing less, but it's at the expense of safety, and makes it harder to guarantee that builds are reproducible.

Concourse does its best to provide these guarantees without sacrificing performance. First, by caching fetched artifacts so that they don't have to be fetched each time a build runs, and are instead fetched at most once-per-worker. Secondly, by creating "copy-on-write" volumes rather than literally copying the data when providing an input or image to a container. This is what the so-called "volume drivers" like `btrfs` and `overlay` are for.

### Privileged and unprivileged volumes

For safety, Concourse runs unprivileged containers with a separate [user namespace](http://man7.org/linux/man-pages/man7/user_namespaces.7.html) that has UID `0` (`root`) mapped to the highest possible UID on the system (say, `4294967294`). This way `root` in the container is actually an unprivileged user on the outside. This is not the case for privileged containers, where `root` in the container is `root` on the host.

This is where things get tricky: the container orchestrator must ensure that the user permissions on each file provided to the container align with its UID namespace. This means we have to be sure to recursively `chown` files coming from a privileged volume from UID `0` to UID `4294967294` before providing them to an unprivileged container, and chown from `4294967294` back to `0` when providing an unprivileged volume to a privileged container.

This presents a question: which format do we keep? Concourse could perhaps cache both, but that would result in more disk usage. From a security standpoint, folks should be using unprivileged containers more often than not, so Concourse optimizes for this by caching the *unprivileged* form of resource caches. That way we usually don't have to perform any `chown` operations - the filesystem is already good to go. The trade-off, obviously, is that when a user *does* need to run a privileged task or resource type, Concourse has to walk through all the files and re-map `4294967294` back to `0`. The performance of this operation is dictated by the volume driver.

## Volume drivers and their trade-offs

Concourse supports three volume drivers:

* `btrfs`, where "copy-on-write" volumes are snapshots of the parent volume
* `overlay`, where "copy-on-write" volumes are an `overlay` mount with the parent volume as the lower directory
* `naive`, where "copy-on-write" volumes are really just a full copy of the parent volume

Darwin and Windows use `naive`, as no alternative has been developed for these platforms.

The Kubernetes Helm chart defaults to `naive` for portability. **This should be the first setting you change.** A warning is printed at startup to bring notice to this.

Otherwise, the default driver will be detected with the following rules:

1. If the underlying filesystem for the worker's `--work-dir` is `btrfs`, then the `btrfs` driver will be used.
1. If `btrfs` is supported by the kernel, the `btrfs` driver will be used.
1. If `overlay` is supported by the kernel, the `overlay` driver will be used.
1. Otherwise, `naive` will be used as a fallback.

You may notice that points 1 and 2 sound redundant; this is actually because we've gone back and forth between `btrfs` and `overlay` as the preference, but if the filesystem is already mounted as `btrfs` that decision will always be made first.

### `btrfs`: fast copy, fast `chown`, "space creep" with loopback, unstable on older kernels

The `btrfs` driver is backed by a "real" filesystem: `btrfs`. It is very fast at creating copy-on-write volumes and performing the `chown` operation necessary for transitioning filesystems between unprivileged and privileged.

The first downside of `btrfs` is that there have been many occurrences of it locking up - either by going "read-only" or putting processes in `D` state forever - and requiring a complete re-create of the worker. The `btrfs` filesystem has been around for a while, but it's still an emerging technology. More recent Kernel versions appear to be more stable, and we don't seem to run into this as often anymore, but it's still something to keep in mind for older systems and something to watch out for in newer ones.

The second downside of `btrfs` is that it may require using a loopback device. If you have a filesystem formatted as `btrfs` that you can provide as the `--work-dir`, great - you can ignore this downside. If not, Concourse will have to create a loopback device with a gigantic sparse file used as the backing store. This approach is fairly portable, but has a couple of downsides:

* Loopback devices are expensive; they are a global system resource, and must be managed carefully. This also makes it really awkward to use in a containerized environment like Kubernetes or Docker, as it's possible to "leak" the loopback allocation if the container is destroyed ungracefully.
* ~~Reported disk usage will never go down. It will eventually plateau, and this can often be treated as nothing more than a metrics anomaly, however there have been reports of suspicious disk usage. I suspect there may be some sort of fragmentation going on from the use of a sparse file.~~ This has been addressed by #5187!

Long story short, "there be dragons" - this loopback device flow is definitely not ideal and not as well understood or supported in the industry.

Side note: I wouldn't really be shocked if the stability issues we've seen are somehow exacerbated by the use of the loopback device and sparse file, but I have no evidence of that or really any concrete reason to think so. Just a gut feeling. In any case I would recommend using a real live `btrfs` filesystem and a recent kernel to have the best chance of avoiding both downsides.

### `overlay`: fast copy, slow `chown`, stable

The `overlay` driver works by creating a mount point using the [`overlay` filesystem](https://www.kernel.org/doc/Documentation/filesystems/overlayfs.txt), with a new and empty upper directory and the parent volume's data as the lower directory. All write operations will go to the upper directory which is "overlayed" on top of the lower directory to form the moint point.

Creating a copy-on-write volume with `overlay` is very fast. Performing the `chown` operation necessary for transitioning filesystems between unprivileged and privileged, however, is very slow, as it involves copying the entire file from the lower directory into the upper directory before changing the permissions.

**This means that tasks resource types with `privileged: true` - including the `docker-image` resource - are much slower to initialize with the `overlay` driver.**

Aside from this (pretty stark) downside, `overlay` is the better choice: it's more portable, it's more stable, and it's otherwise very fast. Unfortunately we've found the performance to be too poor for such a common use case, so we made `btrfs` the default after recent kernels showed improved stability.

This is still a high priority to fix, but it's at the intersection of developing technologies and, well, things just haven't progressed quickly enough. Something like [`shiftfs`](https://lwn.net/Articles/757650/) would really help, but its development pace is inconsistent and sometimes stagnant. New `overlay` features like "metadata only copy up" looked promising but are unsafe for untrusted workloads and require recent kernels with the feature available and enabled.

We are currently in the midst of [transitioning to `containerd`](http://localhost:8000/projects/MDc6UHJvamVjdDMxNTcxMzE=), which will allow us to re-think our policy for privileged vs. unprivileged containers: if we *always* user namespaces for containers, privileged or no, we won't have to do *any* `chown` operations, which would **completely eliminate this problem**. However we will have to make this change carefully, as it has significant implications for things like image building and likely other workflows which require privileged containers.

### `naive`: slow copy, fast `chown`, literally just `cp -rp`

The `naive` driver is tried and true, but it is **extremely inefficient** and should only be used as a last resort. There isn't much more to say here: it literally copies everything. Copy speed is subject to your disk's physical peformance, and space usage obviously high as all copying is frontloaded rather than being deferred to when write operations are performed. The `chown` is fast, I mean, technically, because it doesn't have to "copy up" like `overlay` - but that doesn't really help anyone haha.

## What is my disk space being used for?

Concourse stores all data in volumes, which are kept around until they're removed via the *garbage collector*. The `web` nodes keep track of which volumes should be kept, performing a GC tick every 30 seconds, and the `worker` nodes ask the `web` nodes which volumes to destroy by performing a "sweep" tick every 30 seconds. This is also where Concourse notices volumes which have (somehow) gone missing from the worker, and the `web` node eventually reaps them from its database.

Let's go over the different types of volumes that the garbage collector keeps track of:

* Resource cache volumes containing the fetched "bits" of a version of a resource. Note that different caches are kept for different `params:` values. These volumes are kept through a whitelist system, retaining caches for...:
  * versions which are the current candidate for a job's [`get` step](https://concourse-ci.org/jobs.html#schema.step.get-step.get).
  * versions which were used as a task's [`image_resource:`](https://concourse-ci.org/tasks.html#schema.task.image_resource) in the latest build of a job.
  * versions that were used as the *config* for another resource - i.e. the images used in [`resource_types:`](https://concourse-ci.org/resource-types.html#schema.resource_type).
* [Task caches](https://concourse-ci.org/tasks.html#schema.task.caches) are garbage collected when the task step is removed from its build plan (identified by the task name), when its job is removed from the pipeline, or when its pipeline itself is removed. They are kept while the pipeline or job is paused.
* User artifacts uploaded through [`fly execute`](https://concourse-ci.org/tasks.html#running-tasks) are garbage collected after 12 hours.
* The following volumes are always associated to a container, and are subject to the `chown` operation mentioned above if the original volume's permissions have to be remapped for container's user namespace. These volumes are all garbage collected after the container is garbage-collected. For build containers, this means when the build becomes non-interceptible.
  * The container image is a copy-on-write of the image volume, mounted to `/`.
  * [Task inputs](https://concourse-ci.org/tasks.html#schema.task.inputs) and [`put` step inputs](https://concourse-ci.org/jobs.html#schema.step.put-step.inputs) are copy-on-writes of the original input volume, mounted to the task or resource's working directory.
  * [Task outputs](https://concourse-ci.org/tasks.html#schema.task.outputs) are either empty volumes created for the task to write to, or in the case where they are configured with a `path:` matching an input `path:`, they are actually the input's copy-on-write volume.
* Base resource type images volumes are garbage collected when the version reported by the worker changes - i.e. after upgrading Concourse.

## Troubleshooting high disk usage

Given the above, it's easy to think up scenarios where you may hit "peak" disk usage:

* Your pipeline runs tasks using many different large container images.
* Your pipeline has resources which are on the hefty side (in the gigabytes).
* You have large task caches, maybe left behind in paused, long-forgotten pipelines.
* Users are frequently running `fly execute` and uploading large artifacts.

Even for small pipelines, container images alone can take up a lot of space. We recommend a **minimum of 100GB** for getting started and then adjusting from there if possible or necessary. (As a side note, on many cloud IaaSes disk space directly determines your performance as well, so I would recommend being liberal with it either way; space is cheap, time is expensive.)

If you've never run your own CI system before, it's easy to underestimte how much disk space you'll need, and once you start hitting your limit it's easy to see that the usage isn't going down and think that there's a leak when in fact this is just Concourse caching bits to keep your builds speedy. In practice, bugs with the `web` node's GC logic have been extremely rare; these are almost never the root cause of supposed leaks. The garbage collector is a critical component in a very "hot" code path - any bugs there get noticed pretty quickly.

If you suspect that something is wrong, as a first step please provide your workers with as much disk space as possible in order to see if the usage plateaus or if it really *does* consumes all that it can.

If something *still* looks wrong, the following information is criticial to provide in a bug report:

* `--work-dir` disk size and filesystem type
* Volume driver: `btrfs` or `overlay`
  * If you're using `btrfs`, you may want to read the "pitfalls" section above.
  * If you're using `naive`, that's probably your problem immediately.
* Kernel version
* Concourse version
* Deployment scheme (Kubernetes, Docker, VMs, BOSH)

If you've managed to put together steps to reproduce so that we can confirm the problem ourselves, that's perfect! We can probably really easily diagnose the problem that way.

If not, for there to be any hope of diagnosing the root cause, you'll need to do some digging yourself since we don't have access to your cluster. See if you can identify patterns: *what* is being left on disk? Is it a particular artifact? Can you deduce where it's being created from? Is it coming from a particular step in your pipeline? Be sure to include any pipeline snippets or metrics that may help.

Another data point that may be a little harder to calculate, but will give both yourself and us a better idea of how much usage to expect, is how large your pipeline's "base footprint" is: after one run of your pipeline, with a single worker, how much space is used?

You may also want to check your worker logs for failed deletes, or "failed-" log messages in general. Maybe the worker volume sweeper is stuck because of a local disk failure? It can also help to peek into kernel logs - this is where `btrfs` panics tend to show up.

With that information collected, please submit a [bug report](https://github.com/concourse/concourse/issues/new?assignees=&labels=bug&template=bug-report.md&title=) if you have steps to reproduce the problem or a precise description of a bug you found. If you still need help troubleshooting, please open a [new discussion](https://github.com/concourse/concourse/discussions/new) and we'll all try to get to the bottom of it!

**Comment by skreddy6673** [comment]
@vito This is a  very good explanation. 

We started our concourse instance with 100GI, then increased to 150GI > 300GI and finally settled at 600GI. 
Here are some of the Pain points i observed over the time as a concourse admin.

1) The worker is shared across the teams. So it's hard to identify the team who are consuming most of the resources.
2) A platform with large number of failed pipelines means large number of containers and volumes sitting there until a new build ran. 

The work-around when i see this issue is to manually Delete the POD PVC and kill the POD, which will spin up a fresh POD.

The Land-worker seems to be an option but not all the time.

**Reply by vito** [reply]
For your second point, https://github.com/concourse/concourse/pull/5431 was just merged which will allow you to configure a maximum duration for failed builds. The default will be 5 days.
